Place the output in the expression

4atj opened the pull request:
Placed the output in `Expr` & removed the `literal` field & used a `HashSet` instead of a `HashMap`.

#### Benefits:
- +5-10% performance boost
- Uses less memory if `Num = i32` & `len() % 2 != 0` or if  `Num = i16` & `len() % 4 != 0`  (in 64 bit environments)
- Enables future implementation of chained operators

4atj committed:
Placed the output in Expr & Removed the literal field

JayXon commented on src/main.rs:
```diff
@@ -418,37 +408,36 @@ fn find_unary_expressions(cn: &mut CacheLevel, cache: &Cache, n: usize) {
-            let vec: Vector = Vector::from_slice(input.vec);
-            cn.insert(vec, Expr::variable(i as Literal));
+        for (i, _input) in INPUTS.iter().enumerate() {
+            cn.insert(Expr::variable(i));
```
nit: I think it's better to pass the input vec to `Expr::variable` since it's already available here, instead of looking it up again, but it doesn't really matter since this is only run once for each input, but at least this should be `for i in 0..INPUTS.len()`

4atj replied:
yeah I forgot to change that.

JayXon commented on src/main.rs:
```diff
@@ -418,37 +408,36 @@ fn find_unary_expressions(cn: &mut CacheLevel, cache: &Cache, n: usize) {
         if let Some(m) = (10 as Num).checked_pow(n as u32 - 1) {
             for lit in m..=m.saturating_mul(9).saturating_add(m - 1).min(MAX_LITERAL) {
-                cn.insert(Vector::constant(lit), Expr::literal(lit as Literal));
+                cn.insert(Expr::literal(lit as Num));
```
`lit` is already a `Num`, no need to `as`

JayXon commented on src/vec.rs:
```diff
@@ -5,7 +5,7 @@ use crate::{
 };
 
-#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
```
Any reason we need `Copy`? I think it's better that we have to explicitly clone, so that we don't accidentally make a copy when it's not necessary, `Vector` could be large.

JayXon commented on src/main.rs:
```diff
@@ -418,37 +408,36 @@ fn find_unary_expressions(cn: &mut CacheLevel, cache: &Cache, n: usize) {
     for &lit in LITERALS {
         if positive_integer_length(lit) == n {
-            cn.insert(Vector::constant(lit), Expr::literal(lit as Literal));
+            cn.insert(Expr::literal(lit as Num));
```
`lit` is already a `Num`, no need to `as`

JayXon commented on src/operator.rs:
```diff
@@ -61,7 +62,8 @@ impl Display for Operator {
             Operator::Parens => write!(f, "("),
-            Operator::Literal => write!(f, ""),
+            Operator::Literal |
+            Operator::Variable => write!(f, ""),
```
hmm did you run `cargo fmt`?

JayXon commented on src/expr.rs:
```diff
@@ -70,8 +74,8 @@ impl Expr {
             op: Operator::Parens,
-            literal: 0,
             var_mask: unsafe { *er.as_ptr() }.var_mask,
+            output: unsafe { *er.as_ptr() }.output
```
I'd write it like this:
```rs
        let Self { var_mask, output, .. } = unsafe { *er.as_ptr() };
        Self {
            left: None,
            right: Some(er),
            op: Operator::Parens,
            var_mask,
            output,
        }
```

JayXon commented on src/main.rs:
```diff
@@ -82,30 +82,30 @@ fn save(level: &mut CacheLevel, output: Vector, expr: Expr, n: usize, cache: &Ca
-fn insert_to_level(level: &mut CacheLevel, output: Vector, expr: Expr) {
-    match level.entry(output) {
-        Entry::Occupied(mut e) => {
+fn insert_to_level(level: &mut CacheLevel, expr: Expr) {
```
could you update the rayon version too? there's a call in line 485

4atj replied:
opsy

JayXon committed:
Make Expr and Vector not Copy

Fixed an unnecessary copy in insert_to_level, even more faster now.
Fixed rayon version compile.
A few small tweaks.
Ran `cargo fmt`.

4atj committed:
Merge pull request #3 from JayXon/hashset

Make Expr and Vector not Copy

lynn committed:
Merge branch 'main' into HashSet
